GeoLib/OctTree.h

```diff
@@ -15,116 +15,80 @@
+	/// domain
+	/// (2) ret_pnt is equal to pnt: the point is added to the OctTree
+	/// (3) In case ret_pnt is neither equal to pnt nor equal to nullptr,
+	/// another item within the eps distance is already in the OctTree and the
+	/// pointer to this object is returned.
+	/// @return If the point can be inserted the method returns true, else false.
+	bool addPoint(POINT * pnt, POINT *& ret_pnt);
+
+	/// range query - returns all points inside the range [min[0], max[0]) x
+	/// [min[1], max[1]) x [min[2], max[2])
+	/// @param min
+	/// @param max
+	/// @param pnts
+	template <typename T>
+	void
+	getPointsInRange(T const& min, T const& max, std::vector<POINT*> &pnts) const;
+
+#ifndef NDEBUG
+	MathLib::Point3d const& getLowerLeftCornerPoint() const { return _ll; }
+	MathLib::Point3d const& getUpperRightCornerPoint() const { return _ur; }
+	OctTree<POINT> const* getChild(std::size_t i) const { return _childs[i]; }
+	std::vector<POINT*> const& getPointVector() const { return _pnts; }
+#endif
 
-	static OctTree<POINT>* createOctTree(POINT & ll, POINT & ur, std::size_t max_points_per_node)
-	{
-		const double dx(ur[0] - ll[0]);
-		const double dy(ur[1] - ll[1]);
-		const double dz(ur[2] - ll[2]);
-
-		if (dx >= dy && dx >= dz) {
-			ll[1] -= (dx-dy)/2.0;
-			ur[1] += (dx-dy)/2.0;
-			ll[2] -= (dx-dz)/2.0;
-			ur[2] += (dx-dz)/2.0;
-		} else {
-			if (dy >= dx && dy >= dz) {
-				ll[0] -= (dy-dx)/2.0;
-				ur[0] += (dy-dx)/2.0;
-				ll[2] -= (dy-dz)/2.0;
-				ur[2] += (dy-dz)/2.0;
-			} else {
-				ll[0] -= (dz-dx)/2.0;
-				ur[0] += (dz-dx)/2.0;
-				ll[1] -= (dz-dy)/2.0;
-				ur[1] += (dz-dy)/2.0;
-			}
-		}
-
-		OctTree<POINT>::_max_points_per_node = max_points_per_node;
-		return new OctTree<POINT>(ll, ur);
-	}
-
-	virtual ~OctTree()
-	{
-		for (std::size_t k(0); k < 8; k++)
-			delete _childs[k];
-	}
-
-	/**
-	 * This method adds the given point to the OctTree. If necessary,
-	 * the OctTree will be extended.
-	 * @param pnt the point
-	 * @return If the point can be inserted the method returns true, else false.
-	 */
-	bool addPoint (POINT* pnt)
-	{
-		if ((*pnt)[0] < _ll[0]) return false;
-		if ((*pnt)[0] > _ur[0]) return false;
-		if ((*pnt)[1] < _ll[1]) return false;
-		if ((*pnt)[1] > _ur[1]) return false;
-		if ((*pnt)[2] < _ll[2]) return false;
-		if ((*pnt)[2] > _ur[2]) return false;
-
-		if (!_is_leaf) {
-			for (std::size_t k(0); k < 8; k++) {
-				if (_childs[k]->addPoint (pnt)) {
-					return true;
-				}
-			}
-		}
-
-		// check if point is already in OctTree
-		bool pnt_in_tree (false);
-		for (std::size_t k(0); k < _pnts.size() && !pnt_in_tree; k++) {
-			const double sqr_dist (MathLib::sqrDist( (_pnts[k])->getCoords(), pnt->getCoords() ));
-			if (sqr_dist < std::numeric_limits<double>::epsilon())
-				pnt_in_tree = true;
-		}
-		if (!pnt_in_tree)
-			_pnts.push_back (pnt);
-		else
-			return false;
-
-		if (_pnts.size () > OctTree<POINT>::_max_points_per_node)
-			splitNode ();
-		return true;
-	}
+private:
+	static double _eps;
```

Comment:
Changed from class variable to class member variable. I also changed the handling of the number of maximal points per `OctTree` node from a static variable to a template parameter. So it is possible to have for each `OctTree` it is own value.